### PR TITLE
primitives/sr25519: Enable generating secret keys from ed25519 keys

### PR DESCRIPTION
**Pull request by jberci**

### CODE REVIEW

File: primitives/sr25519/keys.go [hunk 1]
```diff
@@ -252,6 +252,28 @@ func NewSecretKeyFromBytes(b []byte) (*SecretKey, error) {
...
+	if len(b) != SecretKeySize {
+		return nil, fmt.Errorf("sr25519: invalid expanded ed25519 key length")
+	}
+
+	scalar, err := scalarDivideByCofactor(b)
+	if err != nil {
+		return nil, err
+	}
+
+	var nonce [SecretKeyNonceSize]byte
```
Comment (peternose) [inline primitives/sr25519/keys.go:267]:
```suggestion
	var hashPrefix [SecretKeyNonceSize]byte
```

**Review by peternose — COMMENTED** [review]
I'm still against using private keys from scheme A in schemes A and B.

**Comment by kostko** [comment]
> I'm still against using private keys from scheme A in schemes A and B.

This would not be the case, the key would only be used in one scheme, not both. But the derivation process is something that is needed for compatibility with a particular hardware device. Also see related implementation in Rust's [schnorrkel](https://github.com/w3f/schnorrkel/blob/ab3e3d609cd8b9eefbe0333066f698c40fd09582/src/keys.rs#L495-L533) for reference.

**Review by peternose — APPROVED** [review]
If we are following the Rust implementation, then I suppose my comments are redundant.

File: primitives/sr25519/keys_test.go [hunk 2]
```diff
@@ -197,6 +198,36 @@ func TestSecretKey(t *testing.T) {
 			t.Fatalf("sk == sk2")
 		}
 	})
+	t.Run("Ed25519Bytes", func(t *testing.T) {
+		ed25519Bytes, _ := hex.DecodeString("28b0ae221c6bb06856b287f60d7ea0d98552ea5a16db16956849aa371db3eb51fd190cce74df356432b410bd64682309d6dedb27c76845daf388557cbac3ca34")
```
Comment (kostko) [inline primitives/sr25519/keys_test.go:202]: Add a note saying where the test vector is from.

File: primitives/sr25519/keys.go [hunk 3]
```diff
@@ -252,6 +252,34 @@ func NewSecretKeyFromBytes(b []byte) (*SecretKey, error) {
...
+// NewSecretKeyFromEd25519Bytes constructs a SecretKey from the byte representation
+// of an expanded Ed25519 key.
+func NewSecretKeyFromEd25519Bytes(b []byte) (*SecretKey, error) {
+	if len(b) != SecretKeySize {
+		return nil, fmt.Errorf("sr25519: invalid expanded ed25519 key length")
+	}
+
+	// Check the scalar is valid, s == clamp(s)
+	s := b[:SecretKeyScalarSize]
+	if s[0] != (s[0]&0b1111_1000) || (s[31] != ((s[31] & 0b0011_1111) | 0b0100_0000)) {
```
Comment (peternose) [inline primitives/sr25519/keys.go:264]:
```suggestion
	if s[0] & 0b0000_0111 != 0 || s[31] & 0b1100_0000 != 0b0100_0000)) {
```

File: primitives/sr25519/keys.go [hunk 4]
```diff
@@ -252,6 +252,34 @@ func NewSecretKeyFromBytes(b []byte) (*SecretKey, error) {
...
 }
 
+// NewSecretKeyFromEd25519Bytes constructs a SecretKey from the byte representation
+// of an expanded Ed25519 key.
+func NewSecretKeyFromEd25519Bytes(b []byte) (*SecretKey, error) {
+	if len(b) != SecretKeySize {
+		return nil, fmt.Errorf("sr25519: invalid expanded ed25519 key length")
+	}
+
+	// Check the scalar is valid, s == clamp(s)
```
Comment (peternose) [inline primitives/sr25519/keys.go:262]:
```suggestion
	// Check the scalar is clamped.
```

File: primitives/sr25519/keys_test.go [hunk 5]
```diff
@@ -197,6 +198,36 @@ func TestSecretKey(t *testing.T) {
...
+		sk, err := NewSecretKeyFromEd25519Bytes(ed25519Bytes)
+		if err != nil {
+			t.Fatalf("NewSecretKeyFromEd25519Bytes: %v", err)
+		}
+
+		pkBytes, err := sk.PublicKey().MarshalBinary()
```
Comment (peternose) [inline primitives/sr25519/keys_test.go:208]: We can remove this, as we are testing if `PublicKey` method works. Here, it is enough to test that the secret key `sk.key` is ok, as public key is derived from it.

File: primitives/sr25519/keys_test.go (hunk 2 above)
Comment (peternose) [reply primitives/sr25519/keys_test.go:202]: If there are multiple vectors, we could also do a for loop over vector slice:
```
testVectors := []testVector{
     {
        ed25519: "28b0ae221c6bb06856b287f60d7ea0d98552ea5a16db16956849aa371db3eb51fd190cce74df356432b410bd64682309d6dedb27c76845daf388557cbac3ca34"),
        sk: "46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a",
        nonce: "fd190cce74df356432b410bd64682309d6dedb27c76845daf388557cbac3ca34",
    }
    ...
}

for _, v := range testVectors {
}
```
We could also test that we get an error if the size is not ok or if we corrupt the scalar.

File: primitives/sr25519/keys.go [hunk 6]
```diff
@@ -252,6 +252,34 @@ func NewSecretKeyFromBytes(b []byte) (*SecretKey, error) {
...
+		return nil, fmt.Errorf("sr25519: invalid expanded ed25519 key length")
+	}
+
+	// Check the scalar is clamped.
+	s := b[:SecretKeyScalarSize]
+	if (s[0]&0b0000_0111) != 0 || (s[31]&0b1100_0000) != 0b0100_0000 {
+		return nil, fmt.Errorf("sr25519: ed25519 scalar invalid")
+	}
+
+	scalar, err := scalarDivideByCofactor(b[:SecretKeyScalarSize])
```
Comment (peternose) [inline primitives/sr25519/keys.go:268]:
```suggestion
	scalar, err := scalarDivideByCofactor(s)
```

File: primitives/sr25519/keys_test.go [hunk 7]
```diff
@@ -197,6 +198,60 @@ func TestSecretKey(t *testing.T) {
...
+			},
+			{
+				bytes:   "2ab0ae221c6bb06856b287f60d7ea0d98552ea5a16db16956849aa371db3eb51fd190cce74df356432b410bd64682309d6dedb27c76845daf388557cbac3ca34",
+				message: "NewSecretKeyFromEd25519Bytes: scalar not divisible by 8 must fail",
+			},
+		} {
+			decoded := testhelpers.MustUnhex(t, d.bytes)
+			_, err = NewSecretKeyFromEd25519Bytes(decoded)
+			if err == nil {
+				t.Fatalf(d.message)
```
Comment (peternose) [inline primitives/sr25519/keys_test.go:229]: ### 
```suggestion
				t.Errorf(d.message)
```

**Review by peternose — APPROVED** [review]
Looks good to me :+1:

**Review by kostko — APPROVED** [review]

**Comment by matevz** [comment]
Some test vectors obtained directly from the Ledger's derivation for the [test mnemonic](https://github.com/Zondax/ledger-oasis/blob/main/README.md#how-to-prepare-your-development-device):

```
PIN: 5555
Mnemonic: equip will roof matter pink blind book anxiety banner elbow sun young
```

| Algorithm | Number | Public Key | Address |
|----------------|--------------|-----------------|-------------|
| sr25519-adr8 |0 | l+cuboPsOeuY1+kYlROrpmKgiiELmXSw9xl0WEg8cWE= | oasis1qpl4axynedmdrrgrg7dpw3yxc4a8crevr5dkuksl |
| sr25519-adr8 | 1 | VOmOqK/PEyHt3SyR7nH3+SN8OL2MMkIFe+XHzj9Gq70= | oasis1qrlps0l9g607535rgshz8urjvpwzuujmgyjm766h |
| sr25519-adr8 | 2 | fRChHhpO9a3qM+ufMzLG0iHBLUYSmd4y0Q5s//zXdtg= | oasis1qr6j5zx4eyvsfdpe6wslnt7nj9cdhst38czrrrmr |
| sr25519-adr8 | 3 | APOgBQkpM+jClW1+zmLL05cYZ441vypzcMNE6edVvBg= | oasis1qpc3uhcleepuztfwu709mhs5rfz5wvt60cg9w27e |

**Pull request merged**